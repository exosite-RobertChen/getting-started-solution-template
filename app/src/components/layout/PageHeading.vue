<template>
  <div
    class="mt-4 pr-2 pl-6 page-heading"
    :class="{ 'mb-2': $vuetify.breakpoint.xs, 'mb-4': !$vuetify.breakpoint.xs }"
  >
    <h1 class="hidden-sm-and-up headline mt-4">{{ name }}</h1>
    <v-row align="center">
      <h1 class="hidden-xs-only mr-5">{{ name }}</h1>
      <slot />
    </v-row>
  </div>
</template>

<script>
export default {
  props: {
    name: {
      required: true,
      type: String
    }
  }
}
</script>

<style lang="scss">
.page-heading {
  h1 {
    font-weight: 300;
  }

  @media only screen and (min-width: 601px) {
    min-height: 60px;
    display: flex;
    align-items: flex-end;
  }
}
</style>
