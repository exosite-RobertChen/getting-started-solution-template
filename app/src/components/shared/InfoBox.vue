<template>
  <v-row
    :class="{ 'subtitle-1': $vuetify.breakpoint.xs }"
    class="headline fill-height"
    justify="center"
    align="center"
  >
    <v-card
      :class="{
        'px-6': $vuetify.breakpoint.xs,
        'px-10': !$vuetify.breakpoint.xs
      }"
      class="center py-10 text-center elevation-5 border-radius-1"
    >
      <div
        v-for="text in texts"
        :key="text"
        class="mb-2"
        :class="{ 'accent--text': error }"
      >
        {{ text }}
      </div>
      <div v-if="btnText">
        <span>Go ahead and</span>
        <v-btn
          class="ml-3"
          color="accent"
          :small="$vuetify.breakpoint.xs"
          data-qa="InfoBoxButton"
          @click="$emit('action')"
          >{{ btnText }}</v-btn
        >
      </div>
      <div class="mt-6">
        <a
          :href="links.help"
          class="subtitle-1 info-box__link"
          target="_blank"
          primary--text
          >Need help?</a
        >
      </div>
    </v-card>
  </v-row>
</template>

<script>
export default {
  name: 'InfoBox',
  props: {
    texts: {
      type: Array,
      required: true
    },
    btnText: {
      type: String,
      default: ''
    },
    links: {
      type: Object,
      required: true
    },
    error: {
      type: Boolean
    }
  }
}
</script>

<style lang="scss">
.info-box {
  &__link {
    text-decoration: underline;
  }
}
</style>
