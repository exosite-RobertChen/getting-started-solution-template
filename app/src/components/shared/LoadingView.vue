<template>
  <v-row justify="center" align="center">
    <v-col class="text-center">
      <v-progress-circular :size="50" color="primary" indeterminate />
      <div class="mt-6" :class="textColorClass">{{ text }}</div>
    </v-col>
  </v-row>
</template>

<script>
export default {
  props: {
    text: {
      type: String,
      required: false,
      default: ''
    },
    textColorClass: {
      type: String,
      required: false,
      default: ''
    }
  }
}
</script>
