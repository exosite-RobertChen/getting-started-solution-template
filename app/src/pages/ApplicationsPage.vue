<template>
  <v-container style="height: 100%">
    <PageHeading name="Applications">
      <v-spacer class="hidden-xs-only" />
      <v-btn
        color="primary"
        class="mr-6"
        :small="$vuetify.breakpoint.xs"
        @click="openConnectAppDialog"
      >
        <v-icon left>cloud</v-icon>
        <span>Connect App</span>
      </v-btn>
    </PageHeading>

    <!-- main content -->
    <v-row v-if="!errors" class="px-4">
      <v-col
        v-for="app in Object.values(apps)"
        :key="app.id"
        :class="{ 'px-0': $vuetify.breakpoint.xs }"
        class="pa-2"
        cols="12"
        sm="6"
        md="4"
        xl="3"
      >
        <v-card
          class="border-radius-0 app-card"
          color="white darken-1"
          data-qa="AppCard"
        >
          <v-card-title class="subtitle-1 my-0" primary-title>
            {{ app.name | truncateText(25) }}
            <v-icon
              v-if="app.tr_st && app.tr_st < 400"
              color="success"
              class="ml-2"
              @click="openAppDetailsDialog(app)"
              >check</v-icon
            >
            <v-icon
              v-else-if="app.tr_st && app.tr_st >= 400"
              color="error"
              class="ml-2"
              @click="openAppDetailsDialog(app)"
              >close</v-icon
            >
            <v-spacer />
            <v-chip
              v-if="app.default === true"
              color="accent"
              class="white--text mr-2"
              >Auto-add</v-chip
            >
            <v-chip class="mr-2" :color="app.type === 1 ? '#baf1ff' : ''">{{
              app.managed ? 'Managed' : app.type === 1 ? 'Murano' : 'External'
            }}</v-chip>
          </v-card-title>
          <a v-if="app.url" :href="app.url" target="_blank">
            <div class="app-logo">
              <img :src="getAppImageSrc(app)" alt="application logo" />
            </div>
          </a>
          <div v-else class="app-logo">
            <img :src="getAppImageSrc(app)" alt="application logo" />
          </div>
          <v-card-text v-if="app.description">
            <p class="mb-0">
              {{ app.description | truncateText(120) }}
            </p>
          </v-card-text>
          <v-spacer
            :style="{ 'min-height': !app.description ? '17px' : '' }"
          ></v-spacer>
          <v-card-actions class="px-4 pb-4">
            <v-btn
              class="mr-2 ml-1"
              color="primary"
              outlined
              @click="openAppDetailsDialog(app)"
              >Settings</v-btn
            >
            <v-spacer />
            <v-icon
              v-if="!app.managed"
              class="delete-icon"
              data-qa="AppCardDelete"
              text
              @click="openDeleteAppDialog(app)"
              >delete</v-icon
            >
          </v-card-actions>
        </v-card>
      </v-col>
    </v-row>

    <h2 v-if="!errors" class="subtitle-1 my-0 pt-4 pl-3">Connect to ..</h2>
    <v-divider v-if="!errors" />
    <AppCards v-if="!errors" />

    <!-- special cases -->
    <v-row
      v-if="showInfoBox || (appsLoading && !errors)"
      justify="center"
      align="center"
      style="height: 60%"
    >
      <LoadingView v-if="appsLoading && !errors" text="Loading applications" />
      <InfoBox
        v-if="showInfoBox"
        :texts="errors || ['You have no configured apps']"
        :btn-text="!errors ? 'Connect App' : null"
        :links="branding.links"
        :error="!!errors"
        @action="openConnectAppDialog()"
      />
    </v-row>

    <!-- dialogs -->
    <AppDetailsDialog
      v-model="appDetailsDialogOpen"
      :app="selectedApp"
      :branding="branding"
      @action-done="appDetailsSaved"
    />
    <AppTokenDialog
      v-model="appTokenDialogOpen"
      :app-token="appToken"
      :branding="branding"
    />
    <ConnectAppDialog
      v-model="connectAppDialogOpen"
      :apps="apps"
      :selected-app="selectedApp"
      :branding="branding"
      @action-done="openAppTokenDialog"
    />
    <DeleteAppDialog
      v-model="deleteAppDialogOpen"
      :app="appToDelete"
      :branding="branding"
      @action-done="$emit('refetchApps')"
    />
  </v-container>
</template>

<script>
import InfoBox from '@/components/shared/InfoBox'
import AppTokenDialog from '@/components/applications/AppTokenDialog'
import AppDetailsDialog from '@/components/applications/AppDetailsDialog'
import ConnectAppDialog from '@/components/applications/ConnectAppDialog'
import DeleteAppDialog from '@/components/applications/DeleteAppDialog'
import PageHeading from '@/components/layout/PageHeading'
import LoadingView from '@/components/shared/LoadingView'
import AppCards from '@/components/applications/AppCards'

export default {
  name: 'AppSection',
  components: {
    InfoBox,
    AppTokenDialog,
    AppDetailsDialog,
    ConnectAppDialog,
    DeleteAppDialog,
    PageHeading,
    LoadingView,
    AppCards
  },
  props: {
    globalError: {
      type: String,
      required: true
    },
    branding: {
      type: Object,
      required: true
    },
    apps: {
      type: Object,
      required: true
    },
    appsLoading: {
      type: Boolean,
      required: true
    }
  },
  data() {
    return {
      search: '',
      appDetailsDialogOpen: false,
      appTokenDialogOpen: false,
      connectAppDialogOpen: false,
      deleteAppDialogOpen: false,
      localError: '',
      appToDelete: {},
      appToken: '',
      selectedApp: {}
    }
  },
  computed: {
    errors() {
      const result = []
      if (this.globalError) {
        result.push(this.globalError)
      }
      if (this.localError) {
        result.push(this.localError)
      }

      return result.length > 0 ? result : null
    },
    showInfoBox() {
      return this.errors || (this.apps.length < 1 && !this.appsLoading)
    },
    showLoading() {
      return this.appsLoading && !this.showInfoBox
    }
  },
  mounted() {
    this.assignApp()
  },
  methods: {
    openAppDetailsDialog(app) {
      this.selectedApp = app
      this.appDetailsDialogOpen = true
    },
    openAppTokenDialog(result) {
      this.$emit('refetchApps')
      this.appToken = result.data.token
      this.appTokenDialogOpen = true
    },
    openConnectAppDialog() {
      this.connectAppDialogOpen = true
    },
    openDeleteAppDialog(app) {
      this.deleteAppDialogOpen = true
      this.appToDelete = app
    },
    getAppImageSrc(app) {
      if (app.logo || app.l) return app.logo || app.l

      return app.type === 1
        ? 'static/apps/murano.png'
        : 'static/apps/external.png'
    },
    appDetailsSaved() {
      this.$emit('refetchApps')
      this.$emit('openSnackbar', 'success', 'Successfully edited application')
    },
    assignApp() {
      if (this.$route && this.$route.params.select) {
        this.selectedApp = this.$route.params.select
        this.appDetailsDialogOpen = true
      }
    }
  }
}
</script>

<style lang="scss">
.api-result {
  background-color: #777;
  color: white;
}
.border-right {
  border-right: 1px solid rgba(0, 0, 0, 0.12);
}
.v-window {
  height: 100% !important;
}

.app-card {
  height: 100%;
  display: flex !important;
  flex-direction: column;
}

.app-logo {
  height: 150px;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 10px 0;
  img {
    max-height: 100% !important;
    max-width: 80% !important;
  }
}

.app-error {
  margin: 0 !important;
}
</style>
