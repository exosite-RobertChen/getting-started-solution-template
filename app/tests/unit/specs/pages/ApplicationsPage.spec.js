import merge from 'lodash.merge'

import { shallowComponent } from '@/../tests/unit/common'
import branding from '@/../vendor/branding.json'
import ApplicationsPage from '@/pages/ApplicationsPage'
import LoadingView from '@/components/shared/LoadingView'
import InfoBox from '@/components/shared/InfoBox'
import PageHeading from '@/components/layout/PageHeading'

const defaultProps = { sync: false,
  propsData: {
    branding,
    globalError: '',
    localError: '',
    apps: {},
    appsLoading: true,
    selectedApp: '',
    appDetailsDialogOpen: false
  },
  computed:{
    showInfoBox() {
       return !this.appsLoading
    }
  }
}

const factory = overrides =>
  shallowComponent(ApplicationsPage, merge(defaultProps, overrides))

describe('ApplicationsPage.vue', () => {
  describe('when apps are loading', () => {
    const wrapper = factory()

    test('can be mounted', () => {
      expect(wrapper.contains(ApplicationsPage)).toBe(true)
    })

    test('renders only heading and spinner initially', () => {
      expect(wrapper.contains(LoadingView)).toBe(true)
      expect(wrapper.contains(PageHeading)).toBe(true)
      expect(wrapper.contains(InfoBox)).toBe(false)
      expect(wrapper.contains('[data-qa=AppCard]')).toBe(false)
    })

    test('when groups has no apps shows an InfoBox', async () => {
      expect(wrapper.contains(InfoBox)).toBe(false)

      wrapper.setProps({
        appsLoading: false,
        showInfoBox: true
      })
      wrapper.vm.$forceUpdate()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.appsLoading).toBe(false)
      expect(wrapper.contains('[data-qa=AppCard]')).toBe(false)
      expect(wrapper.contains(LoadingView)).toBe(false)
      expect(wrapper.contains(InfoBox)).toBe(true)
    })
  })

  describe('when there are apps loaded', () => {
    let wrapper

    beforeEach(() => {
      wrapper = factory()

      wrapper.setMethods({ fetchApps: jest.fn() })
      wrapper.setProps({
        appsLoading: false,
        apps: {
          adsfhgdas: {
            name: 'app1',
            tr_st: 200,
            default: true,
            online: true,
            type: 1
          },
          myAwesomeId: {
            name: 'app2',
            tr_st: 401,
            default: false,
            type: 1,
            description:
              'dfasilhgads flahksfd aslhdkflashkd flakhsdflkah sdlfhas dhaoighsh-ph3-qhrreoghaobh asd'
          }
        }
      })
    })

    afterEach(() => {
      wrapper.destroy()
    })

    test('renders an app card each app', () => {
      expect(wrapper.findAll('[data-qa=AppCard]').length).toBe(2)
    })

    test('matches snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    test('opens delete app dialog on when deleteAppDialogOpen called', async () => {
      expect(wrapper.vm.deleteAppDialogOpen).toBe(false)
      wrapper.vm.openDeleteAppDialog({
        aid: 'app1',
        name: 'appName1',
        online: true,
        type: 1
      })
      expect(wrapper.vm.deleteAppDialogOpen).toBe(true)
    })

    test('1st app with error status show second icon, error color', () => {
      expect(wrapper.findAll('.ml-2').at(0).attributes('color')).toBe('success')
    })
    
    test('2nd app with error status show second icon, error color', () => {
      expect(wrapper.findAll('.ml-2').at(1).attributes('color')).toBe('error')
    })
  })
})
    
