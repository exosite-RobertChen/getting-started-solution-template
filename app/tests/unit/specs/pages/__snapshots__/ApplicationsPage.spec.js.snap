// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationsPage.vue when there are apps loaded matches snapshot 1`] = `
<v-container-stub
  localerror=""
  selectedapp=""
  tag="div"
>
  <pageheading-stub
    name="Applications"
  >
    <v-spacer-stub
      class="hidden-xs-only"
    />
     
    <v-btn-stub
      activeclass=""
      class="mr-6"
      color="primary"
      tag="button"
      type="button"
    >
      <v-icon-stub
        left=""
      >
        cloud
      </v-icon-stub>
       
      <span>
        Connect App
      </span>
    </v-btn-stub>
  </pageheading-stub>
   
  <v-row-stub
    class="px-4"
    tag="div"
  >
    <v-col-stub
      class="pa-2 [object Object]"
      cols="12"
      md="4"
      sm="6"
      tag="div"
      xl="3"
    >
      <v-card-stub
        class="border-radius-0 app-card"
        color="white darken-1"
        data-qa="AppCard"
        loaderheight="4"
        tag="div"
      >
        <v-card-title-stub
          class="subtitle-1 my-0"
          primary-title=""
          primarytitle=""
        >
          
          app1
          
          <v-icon-stub
            class="ml-2"
            color="success"
          >
            check
          </v-icon-stub>
           
          <v-spacer-stub />
           
          <v-chip-stub
            active="true"
            activeclass=""
            class="white--text mr-2"
            closeicon="$delete"
            color="accent"
            filtericon="$complete"
            tag="span"
          >
            Auto-add
          </v-chip-stub>
           
          <v-chip-stub
            active="true"
            activeclass=""
            class="mr-2"
            closeicon="$delete"
            color="#baf1ff"
            filtericon="$complete"
            tag="span"
          >
            Murano
          </v-chip-stub>
        </v-card-title-stub>
         
        <div
          class="app-logo"
        >
          <img
            alt="application logo"
            src="static/apps/murano.png"
          />
        </div>
         
        <!---->
         
        <v-spacer-stub />
         
        <v-card-actions-stub
          class="px-4 pb-4"
        >
          <v-btn-stub
            activeclass=""
            class="mr-2 ml-1"
            color="primary"
            outlined="true"
            tag="button"
            type="button"
          >
            Settings
          </v-btn-stub>
           
          <v-spacer-stub />
           
          <v-icon-stub
            class="delete-icon"
            data-qa="AppCardDelete"
            dataqa="AppCardDelete"
            text=""
          >
            delete
          </v-icon-stub>
        </v-card-actions-stub>
      </v-card-stub>
    </v-col-stub>
    <v-col-stub
      class="pa-2 [object Object]"
      cols="12"
      md="4"
      sm="6"
      tag="div"
      xl="3"
    >
      <v-card-stub
        class="border-radius-0 app-card"
        color="white darken-1"
        data-qa="AppCard"
        loaderheight="4"
        tag="div"
      >
        <v-card-title-stub
          class="subtitle-1 my-0"
          primary-title=""
          primarytitle=""
        >
          
          app2
          
          <v-icon-stub
            class="ml-2"
            color="error"
          >
            close
          </v-icon-stub>
           
          <v-spacer-stub />
           
          <!---->
           
          <v-chip-stub
            active="true"
            activeclass=""
            class="mr-2"
            closeicon="$delete"
            color="#baf1ff"
            filtericon="$complete"
            tag="span"
          >
            Murano
          </v-chip-stub>
        </v-card-title-stub>
         
        <div
          class="app-logo"
        >
          <img
            alt="application logo"
            src="static/apps/murano.png"
          />
        </div>
         
        <v-card-text-stub>
          <p
            class="mb-0"
          >
            
            dfasilhgads flahksfd aslhdkflashkd flakhsdflkah sdlfhas dhaoighsh-ph3-qhrreoghaobh asd
          
          </p>
        </v-card-text-stub>
         
        <v-spacer-stub />
         
        <v-card-actions-stub
          class="px-4 pb-4"
        >
          <v-btn-stub
            activeclass=""
            class="mr-2 ml-1"
            color="primary"
            outlined="true"
            tag="button"
            type="button"
          >
            Settings
          </v-btn-stub>
           
          <v-spacer-stub />
           
          <v-icon-stub
            class="delete-icon"
            data-qa="AppCardDelete"
            dataqa="AppCardDelete"
            text=""
          >
            delete
          </v-icon-stub>
        </v-card-actions-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
   
  <h2
    class="subtitle-1 my-0 pt-4 pl-3"
  >
    Connect to ..
  </h2>
   
  <v-divider-stub />
   
  <appcards-stub />
   
  <v-row-stub
    align="center"
    justify="center"
    tag="div"
  >
    <!---->
     
    <infobox-stub
      btntext="Connect App"
      links="[object Object]"
      texts="You have no configured apps"
    />
  </v-row-stub>
   
  <appdetailsdialog-stub
    app="[object Object]"
    branding="[object Object]"
  />
   
  <apptokendialog-stub
    apptoken=""
    branding="[object Object]"
  />
   
  <connectappdialog-stub
    apps="[object Object]"
    branding="[object Object]"
    selected-app="[object Object]"
  />
   
  <deleteappdialog-stub
    app="[object Object]"
    branding="[object Object]"
  />
</v-container-stub>
`;
